File: racer.cs
// racer.cs - Гоночная игра в консоли (C#)
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading;
using System.Linq;

class Racer
{
    private const int WIDTH = 11;
    private const int HEIGHT = 15;
    private const int ROAD_LEFT = 2;
    private const int ROAD_RIGHT = WIDTH - 3;
    private const char CAR = '▲';
    private const char OBSTACLE = '█';
    private const string RECORD_FILE = "racer_record.json";

    private int playerPos;
    private List<(int x, int y)> obstacles;
    private int score;
    private int speed;
    private bool gameOver;
    private bool paused;
    private bool running;
    private int record;
    private Random rand;

    public Racer()
    {
        playerPos = WIDTH / 2;
        obstacles = new List<(int, int)>();
        score = 0;
        speed = 1;
        gameOver = false;
        paused = false;
        running = true;
        record = LoadRecord();
        rand = new Random();
    }

    private int LoadRecord()
    {
        if (!File.Exists(RECORD_FILE)) return 0;
        string json = File.ReadAllText(RECORD_FILE);
        var data = JsonSerializer.Deserialize<Dictionary<string, int>>(json);
        return data != null && data.ContainsKey("record") ? data["record"] : 0;
    }

    private void SaveRecord()
    {
        var data = new Dictionary<string, int> { { "record", record } };
        string json = JsonSerializer.Serialize(data);
        File.WriteAllText(RECORD_FILE, json);
    }

    private void SpawnObstacle()
    {
        int x = rand.Next(ROAD_LEFT, ROAD_RIGHT + 1);
        obstacles.Add((x, 0));
    }

    private void Update()
    {
        if (gameOver || paused) return;

        // Сдвиг вниз
        for (int i = obstacles.Count - 1; i >= 0; i--)
        {
            var obs = obstacles[i];
            obs.y++;
            obstacles[i] = obs;
            if (obs.y >= HEIGHT) obstacles.RemoveAt(i);
        }

        // Столкновение
        foreach (var obs in obstacles)
        {
            if (obs.y == HEIGHT - 1 && obs.x == playerPos)
            {
                gameOver = true;
                if (score > record)
                {
                    record = score;
                    SaveRecord();
                }
                return;
            }
        }

        score++;
        speed = 1 + score / 10;

        // Спавн
        double prob = 0.3 + Math.Min(0.4, score / 200.0);
        if (rand.NextDouble() < prob)
            SpawnObstacle();
    }

    private void Draw()
    {
        Console.Clear();
        Console.WriteLine(new string('═', WIDTH));
        Console.WriteLine($"  Счёт: {score}   Рекорд: {record}   Скорость: {speed}");
        Console.WriteLine(new string('═', WIDTH));

        for (int y = 0; y < HEIGHT; y++)
        {
            for (int x = 0; x < WIDTH; x++)
            {
                if (x == 0 || x == WIDTH - 1)
                    Console.Write('│');
                else
                {
                    bool isObs = obstacles.Any(o => o.x == x && o.y == y);
                    if (isObs)
                        Console.Write(OBSTACLE);
                    else if (y == HEIGHT - 1 && x == playerPos)
                        Console.Write(CAR);
                    else
                        Console.Write(' ');
                }
            }
            Console.WriteLine();
        }
        Console.WriteLine(new string('═', WIDTH));
        Console.WriteLine($"  {(paused ? "ПАУЗА" : "ИГРА")}  |  ← → движение  |  P - пауза  |  R - рестарт  |  Q - выход");
    }

    private void Reset()
    {
        playerPos = WIDTH / 2;
        obstacles.Clear();
        score = 0;
        speed = 1;
        gameOver = false;
        paused = false;
    }

    public void Run()
    {
        DateTime lastUpdate = DateTime.Now;
        while (running)
        {
            // Обработка ввода (неблокирующая)
            while (Console.KeyAvailable)
            {
                var key = Console.ReadKey(true);
                switch (key.Key)
                {
                    case ConsoleKey.LeftArrow:
                    case ConsoleKey.A:
                        if (!gameOver && !paused) playerPos = Math.Max(ROAD_LEFT, playerPos - 1);
                        break;
                    case ConsoleKey.RightArrow:
                    case ConsoleKey.D:
                        if (!gameOver && !paused) playerPos = Math.Min(ROAD_RIGHT, playerPos + 1);
                        break;
                    case ConsoleKey.P:
                        paused = !paused;
                        break;
                    case ConsoleKey.R:
                        Reset();
                        break;
                    case ConsoleKey.Q:
                        running = false;
                        break;
                }
            }

            // Обновление по таймеру
            var now = DateTime.Now;
            if ((now - lastUpdate).TotalSeconds >= 1.0 / speed)
            {
                Update();
                Draw();
                lastUpdate = now;
            }
            Thread.Sleep(20);
        }
    }

    static void Main()
    {
        var game = new Racer();
        game.Run();
        Console.WriteLine("Игра завершена.");
    }
}

File: Racer.java
// Racer.java - Гоночная игра в консоли (Java)
import org.jline.terminal.Terminal;
import org.jline.terminal.TerminalBuilder;
import org.jline.keymap.KeyMap;
import org.jline.reader.LineReader;
import org.jline.reader.LineReaderBuilder;
import org.jline.reader.UserInterruptException;
import org.jline.utils.InfoCmp;

import java.io.IOException;
import java.util.*;
import java.nio.file.*;
import com.google.gson.*;

public class Racer {
    private static final int WIDTH = 11;
    private static final int HEIGHT = 15;
    private static final int ROAD_LEFT = 2;
    private static final int ROAD_RIGHT = WIDTH - 3;
    private static final char CAR = '▲';
    private static final char OBSTACLE = '█';
    private static final String RECORD_FILE = "racer_record.json";

    private Terminal terminal;
    private int playerPos;
    private List<int[]> obstacles; // [x, y]
    private int score;
    private int speed;
    private boolean gameOver;
    private boolean paused;
    private int record;
    private boolean running;
    private long lastUpdate;

    public Racer() throws IOException {
        terminal = TerminalBuilder.builder().system(true).build();
        playerPos = WIDTH / 2;
        obstacles = new ArrayList<>();
        score = 0;
        speed = 1;
        gameOver = false;
        paused = false;
        running = true;
        record = loadRecord();
        lastUpdate = System.currentTimeMillis();
    }

    private int loadRecord() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(RECORD_FILE)));
            JsonObject obj = new Gson().fromJson(content, JsonObject.class);
            return obj.get("record").getAsInt();
        } catch (Exception e) {
            return 0;
        }
    }

    private void saveRecord() {
        try {
            JsonObject obj = new JsonObject();
            obj.addProperty("record", record);
            Files.write(Paths.get(RECORD_FILE), obj.toString().getBytes());
        } catch (Exception e) {}
    }

    private void spawnObstacle() {
        int x = ROAD_LEFT + (int)(Math.random() * (ROAD_RIGHT - ROAD_LEFT + 1));
        obstacles.add(new int[]{x, 0});
    }

    private void update() {
        if (gameOver || paused) return;

        // Сдвиг вниз
        for (int[] obs : obstacles) {
            obs[1]++;
        }
        obstacles.removeIf(obs -> obs[1] >= HEIGHT);

        // Столкновение
        for (int[] obs : obstacles) {
            if (obs[1] == HEIGHT - 1 && obs[0] == playerPos) {
                gameOver = true;
                if (score > record) {
                    record = score;
                    saveRecord();
                }
                return;
            }
        }
        score++;
        speed = 1 + score / 10;

        // Спавн
        if (Math.random() < 0.3 + Math.min(0.4, score / 200.0)) {
            spawnObstacle();
        }
    }

    private void draw() {
        terminal.puts(InfoCmp.Capability.clear_screen);
        System.out.println("═".repeat(WIDTH));
        System.out.printf("  Счёт: %d   Рекорд: %d   Скорость: %d%n", score, record, speed);
        System.out.println("═".repeat(WIDTH));

        for (int y = 0; y < HEIGHT; y++) {
            StringBuilder line = new StringBuilder();
            for (int x = 0; x < WIDTH; x++) {
                if (x == 0 || x == WIDTH - 1) {
                    line.append('│');
                } else {
                    boolean isObs = false;
                    for (int[] obs : obstacles) {
                        if (obs[0] == x && obs[1] == y) {
                            isObs = true;
                            break;
                        }
                    }
                    if (isObs) {
                        line.append(OBSTACLE);
                    } else if (y == HEIGHT - 1 && x == playerPos) {
                        line.append(CAR);
                    } else {
                        line.append(' ');
                    }
                }
            }
            System.out.println(line);
        }
        System.out.println("═".repeat(WIDTH));
        String status = paused ? "ПАУЗА" : "ИГРА";
        System.out.printf("  %s  |  ← → движение  |  P - пауза  |  R - рестарт  |  Q - выход%n", status);
    }

    private void reset() {
        playerPos = WIDTH / 2;
        obstacles.clear();
        score = 0;
        speed = 1;
        gameOver = false;
        paused = false;
    }

    private void handleInput() {
        try {
            while (running) {
                int ch = terminal.reader().read();
                if (ch == -1) continue;
                char c = (char) ch;
                switch (c) {
                    case 'a': case 'A': case 27: // left arrow: ESC [ D
                        // обработаем отдельно
                        break;
                    default:
                        break;
                }
                // Для стрелок используем терминальный ввод
                // В jline сложно обрабатывать стрелки в цикле, поэтому используем KeyMap
                // Упростим: используем обычные клавиши
                if (c == 'a' || c == 'A' || c == '←') {
                    if (!gameOver && !paused) playerPos = Math.max(ROAD_LEFT, playerPos - 1);
                } else if (c == 'd' || c == 'D' || c == '→') {
                    if (!gameOver && !paused) playerPos = Math.min(ROAD_RIGHT, playerPos + 1);
                } else if (c == 'p' || c == 'P') {
                    paused = !paused;
                } else if (c == 'r' || c == 'R') {
                    reset();
                } else if (c == 'q' || c == 'Q') {
                    running = false;
                    break;
                }
                // Для стрелок: в jline читаем последовательности, но для упрощения оставим буквы
                // В полноценной версии использовать KeyMap
                // Здесь для демонстрации используем a/d
            }
        } catch (IOException e) {}
    }

    public void run() throws InterruptedException {
        Thread inputThread = new Thread(this::handleInput);
        inputThread.setDaemon(true);
        inputThread.start();

        while (running) {
            long now = System.currentTimeMillis();
            if (now - lastUpdate >= 1000 / speed) {
                update();
                draw();
                lastUpdate = now;
            }
            Thread.sleep(50);
        }
        terminal.close();
    }

    public static void main(String[] args) throws Exception {
        Racer game = new Racer();
        game.run();
    }
}
